### PR TITLE
Generate failureDomains from availabilityZones for CAPC

### PR DESCRIPTION
**Pull request by wongni**
*Issue #, if available:*
https://github.com/aws/eks-anywhere/issues/2406

*Description of changes:*
This PR uses CAPC v1beta2 and generate failureDomains instead of Zones to support mutiple ACS endpoints. 

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_types.go [hunk 1]
```diff
@@ -177,7 +177,7 @@ func (v *CloudStackDatacenterConfig) SetDefaults() {
 				Account:               v.Spec.Account,
 				Domain:                v.Spec.Domain,
 				ManagementApiEndpoint: v.Spec.ManagementApiEndpoint,
-				CredentialsRef:        "Global",
+				CredentialsRef:        "global",
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_types.go:180]: let's also resolve the TODO above in the Validate() method. I think we can call Validate() from cloudstack.go or validator.go

File: controllers/resource/fetcher.go [hunk 2]
```diff
@@ -589,20 +589,22 @@ func MapMachineTemplateToVSphereMachineConfigSpec(vsMachineTemplate *vspherev1.V
...
-	var zones []anywherev1.CloudStackZone
-	for _, csZone := range csCluster.Spec.Zones {
-		zones = append(zones, anywherev1.CloudStackZone{
-			Id:   csZone.ID,
-			Name: csZone.Name,
-			Network: anywherev1.CloudStackResourceIdentifier{
-				Id:   csZone.Network.ID,
-				Name: csZone.Network.Name,
+	var azs []anywherev1.CloudStackAvailabilityZone
+	for _, fd := range csCluster.Spec.FailureDomains {
```
Comment (maxdrib) [inline controllers/resource/fetcher.go:593]: I think we've answered this before - but will there be any point at which this updated eks-a controller is reading in v1beta1 CAPC resources, and has to map them back to AZ's? This code assumes we always have FailureDomains present

File: go.mod [hunk 3]
```diff
@@ -185,4 +185,5 @@ replace (
 
 	// need the modifications eksa made to the capi api structs
 	sigs.k8s.io/cluster-api => github.com/mrajashree/cluster-api v1.1.3-custom
+	sigs.k8s.io/cluster-api-provider-cloudstack => github.com/rejoshed/cluster-api-provider-cloudstack v0.4.5-rc3.0.20220714221707-f31eecbdff3a
```
Comment (maxdrib) [inline go.mod:188]: nice :) maybe add a comment explaining why this is needed

File: pkg/api/v1alpha1/cloudstackdatacenterconfig_types.go [hunk 4]
```diff
@@ -160,10 +161,29 @@ func (v *CloudStackDatacenterConfig) Marshallable() Marshallable {
...
+	if len(v.Spec.Zones) > 0 {
+		return errors.New("zones must be empty")
+	}
+	if len(v.Spec.AvailabilityZones) == 0 {
+		return errors.New("availabilityZones must not be empty")
+	}
+	azSet := map[string]CloudStackAvailabilityZone{}
+	for _, az := range v.Spec.AvailabilityZones {
+		if _, exists := azSet[az.Name]; exists {
+			return errors.New(fmt.Sprintf("All availabilityZone names must not be unique; found multiple %s", az.Name))
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackdatacenterconfig_types.go:182]:
```suggestion
			return errors.New(fmt.Sprintf("All availabilityZone names must be unique; found multiple %s", az.Name))
```

File: pkg/providers/cloudstack/config/template-cp.yaml [hunk 5]
```diff
@@ -35,25 +35,19 @@ spec:
...
+  failureDomains:{{ range $az := .cloudstackAvailabilityZones }}
+  - name: {{ $az.Name }}
+    zone:
+      name: {{ $az.Zone.Name }}
+      network:
+        name: {{ $az.Zone.Network.Name }}
+    domain: {{ $az.Domain }}
+    account: {{ $az.Account }}
+    acsendpoint:
+      name: {{ $az.CredentialsRef }}
```
Comment (maxdrib) [inline pkg/providers/cloudstack/config/template-cp.yaml:47]: do we need a namespace here in addition to the credentialRef name?

**Review by maxdrib — COMMENTED** [review]
I'd be curious to see how this was tested e2e. Does it work e2e? Is the branch of CAPC ready to roll? Would you be able to run the TestCloudStackKubernetes120UpgradeFromLatestMinorRelease test with this configuration?

Great work so far on all this, by the way. It's all so complicated but you've managed to split it up simply and effectively.

File: pkg/providers/cloudstack/config/template-cp.yaml (hunk 5 above)
Comment (wongni) [reply pkg/providers/cloudstack/config/template-cp.yaml:47]: Yes, capc needs both name and namespace to get a secret. If we don't pass the namespace capc has to have a hardcoded namespace `eksa-system`.

File: pkg/providers/cloudstack/testdata/cluster_availability_zones.yaml [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+spec:
+  availabilityZones:
+  - name: az-1
+    domain: domain-1
+    account: account-1
+    zone:
+      name: zone-1
+      network:
+        name: network-1
+    credentialsRef: global
```
Comment (maxdrib) [inline pkg/providers/cloudstack/testdata/cluster_availability_zones.yaml:51]: could you use different credentials and configurations for each az?

**Review by maxdrib — COMMENTED** [review]
Code LGTM, just the ask for the test case. Could you also make a note about how/when this PR could be merged, and the CAPC dependency we are waiting on?

**Comment by wongni** [comment]
Splited into https://github.com/aws/eks-anywhere/pull/2766, https://github.com/wongni/eks-anywhere/pull/7, and https://github.com/wongni/eks-anywhere/pull/8

**Pull request closed without merging**